Tasks/Maven/mavencodeanalysis.ts

```diff
@@ -0,0 +1,231 @@
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+/// <reference path="../../definitions/Q.d.ts" />
+
+import path = require('path');
+import fs = require('fs');
+
+import tl = require('vsts-task-lib/task');
+import trm = require('vsts-task-lib/toolrunner');
+
+// Lowercased names are to lessen the likelihood of xplat issues
+import pmd = require('./pmdformaven');
+import ar = require('./analysisresult');
+import ma = require('./moduleanalysis');
+
+// Set up for localization
+tl.setResourcePath(path.join( __dirname, 'task.json'));
+
+// Cache build variables - if they are null, we are in a test env and can use test inputs
+// The artifact staging directory will be a subdirectory just to be safe.
+var sourcesDir:string = tl.getVariable('build.sourcesDirectory') || tl.getInput('test.sourcesDirectory');
+var stagingDir:string = path.join(tl.getVariable('build.artifactStagingDirectory') || tl.getInput('test.artifactStagingDirectory'), ".codeAnalysis");
+var buildNumber:string = tl.getVariable('build.buildNumber') || tl.getInput('test.buildNumber');
+
+// Returns the names of any enabled code analysis tools, or empty array if none.
+function getEnabledCodeAnalysisTools():string[] {
+    var result:string[] = [];
+
+    if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
+        console.log('PMD analysis is enabled');
+        result.push(pmd.toolName);
+    }
+
+    return result;
+}
+
+// Returns the full path of the staging directory for a given tool.
+function getStagingDirectory(toolName:string):string {
+    return path.join(stagingDir, toolName.toLowerCase());
+}
+
+function cleanDirectory(targetDirectory:string):boolean {
+    tl.rmRF(targetDirectory);
+    tl.mkdirP(targetDirectory);
+
+    return tl.exist(targetDirectory);
+}
+
+// Apply goals for enabled code analysis tools
+export function applyEnabledCodeAnalysisGoals(mvnRun: trm.ToolRunner):void {
+    var enabledCodeAnalysisTools:string[] = getEnabledCodeAnalysisTools();
+
+    // PMD
+    if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
+        pmd.applyPmdArgs(mvnRun);
+    }
+}
```

Comment:
This method has been altered to be more general and have only a very small PMD-specific section, in line with the rest of the design.